File: leetcode/remove-duplicates-from-sorted-list-ii.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
 public:
    ListNode* deleteDuplicates(ListNode* head)
    {
        ListNode* remover = nullptr;
        ListNode* new_head = nullptr;
        ListNode* now = head;
        ListNode* next;

        while (now != nullptr) {
            if (now->next != nullptr && now->next->val == now->val) {
                next = now->next;
                while (next != nullptr && next->val == now->val) {
                    next = next->next;
                }
                now = next;
            } else {
                if (remover == nullptr) {
                    new_head = now;
                    remover = new_head;
                } else {
                    remover->next = now;
                    remover = now;
                }
                now = now->next;
            }
        }
        if (remover) {
            remover->next = nullptr;
        }

        return new_head;
    }
};

File: leetcode/remove-duplicates-from-sorted-list-ii.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)

        current = dummy
        ptr = head
        next = head.next if head else None

        while next:
            if ptr.val == next.val:
                while next and ptr.val == next.val:
                    ptr = next
                    next = next.next
                current.next = next
            else:
                current.next = ptr
                current = ptr

            ptr = next
            next = next.next if next else None

        return dummy.next
